In [1090]:
 #!/usr/bin/env python

# Lesson 4 Assignment

In [1091]:
import pandas as pd

## Question 1

1. Explore the datasets. (10 points)

In [1092]:
data = pd.read_csv('Assignment4_data.csv')

In [1093]:
data

one   two  three  four  five variable
0   -92.0 -76.0  -33.0   3.0 -13.0       B2
1   -21.0  76.0   38.0  -6.0  80.0       B1
2    -2.0 -47.0  -34.0 -86.0 -66.0       A1
3   -76.0  43.0    7.0 -40.0 -42.0       A1
4    44.0  37.0   -7.0 -14.0  30.0       A1
..    ...   ...    ...   ...   ...      ...
195  63.0   3.0  -30.0 -24.0 -59.0       A1
196  97.0 -48.0  -61.0 -25.0 -21.0       B1
197 -93.0 -75.0  -18.0 -67.0 -58.0       B1
198  54.0 -66.0  -80.0  92.0  62.0       A1
199  82.0  53.0  -77.0  79.0  97.0       B2

[200 rows x 6 columns]

In [1094]:
data.describe()

one         two       three        four        five
count  195.000000  197.000000  199.000000  194.000000  196.000000
mean    -2.656410    2.208122    2.095477   -2.829897   -2.612245
std     67.489135   53.116759  101.864120   87.098996   84.158719
min   -363.000000 -100.000000 -100.000000 -576.000000 -821.000000
25%    -54.500000  -44.000000  -60.000000  -52.500000  -42.250000
50%      0.000000   -1.000000   -9.000000   -8.000000   -1.000000
75%     52.000000   45.000000   45.000000   44.000000   54.250000
max     97.000000   97.000000  832.000000  728.000000   99.000000

In [1095]:
data.corr()

/var/folders/lw/3zvwvsp54zsg427dstxg7gq00000gn/T/ipykernel_30614/2627137660.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.corr()


one       two     three      four      five
one    1.000000  0.026397  0.013446 -0.026705 -0.070970
two    0.026397  1.000000  0.009702  0.003254  0.174418
three  0.013446  0.009702  1.000000  0.072565 -0.014384
four  -0.026705  0.003254  0.072565  1.000000 -0.046798
five  -0.070970  0.174418 -0.014384 -0.046798  1.000000

In [1096]:
data.cov()

/var/folders/lw/3zvwvsp54zsg427dstxg7gq00000gn/T/ipykernel_30614/4020818503.py:1: FutureWarning: The default value of numeric_only in DataFrame.cov is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.cov()


one          two         three         four         five
one    4554.783399    95.107821     93.086961  -158.365417  -405.657840
two      95.107821  2821.390138     52.918001    15.158942   787.776473
three    93.086961    52.918001  10376.298919   651.406439  -124.510838
four   -158.365417    15.158942    651.406439  7586.235164  -277.775439
five   -405.657840   787.776473   -124.510838  -277.775439  7082.689901

The count of each column tells me there are some missing values in the data. Examining the quartiles shows the wide range and high variance of the values in the data. Each of the columns seem to have very little relation to one another. Although the relationship is weak, columns `two` and `five` have the strongest relationship amongst all other columns.

2. Find and handle missing values in the data. (It is your choice how you handle the missing data.) ( 20 points)

In [1097]:
data.isnull().sum(axis=0)

one         5
two         3
three       1
four        6
five        4
variable    0
dtype: int64

To identify any missing values, I counted the number of missing values in each column. There were a total of 19 missing values in the entire dataset.

In [1098]:
data.dropna(how='all')


one   two  three  four  five variable
0   -92.0 -76.0  -33.0   3.0 -13.0       B2
1   -21.0  76.0   38.0  -6.0  80.0       B1
2    -2.0 -47.0  -34.0 -86.0 -66.0       A1
3   -76.0  43.0    7.0 -40.0 -42.0       A1
4    44.0  37.0   -7.0 -14.0  30.0       A1
..    ...   ...    ...   ...   ...      ...
195  63.0   3.0  -30.0 -24.0 -59.0       A1
196  97.0 -48.0  -61.0 -25.0 -21.0       B1
197 -93.0 -75.0  -18.0 -67.0 -58.0       B1
198  54.0 -66.0  -80.0  92.0  62.0       A1
199  82.0  53.0  -77.0  79.0  97.0       B2

[200 rows x 6 columns]

First, I decided that any row that had data missing in all columns should be dropped. There didnt appear to be any such row so I used another approach. Given the large variance of each column, I didnt feel it was appropriate to replace all missing values with its column's mean or zero. Instead, I chose to use fill any missing values with the value at the same column but previous row. 

In [1099]:
data = data.fillna(method='ffill')

In [1100]:
data.isnull().sum(axis=0)

one         0
two         0
three       0
four        0
five        0
variable    0
dtype: int64

Now, there are no more missing values!

3. Explore the variable column and convert the "variable” column to dummy variables and join the dummies to the data. (20 points)

In [1101]:
data.variable.value_counts()

A1    59
B2    50
A2    46
B1    45
Name: variable, dtype: int64

We see there are four unique values in the `variable` column and there is a fairly even distribution of rows with each category.

In [1102]:
dummies = pd.get_dummies(data.variable)

In [1103]:
dummy_data = data.join(dummies)
dummy_data

one   two  three  four  five variable  A1  A2  B1  B2
0   -92.0 -76.0  -33.0   3.0 -13.0       B2   0   0   0   1
1   -21.0  76.0   38.0  -6.0  80.0       B1   0   0   1   0
2    -2.0 -47.0  -34.0 -86.0 -66.0       A1   1   0   0   0
3   -76.0  43.0    7.0 -40.0 -42.0       A1   1   0   0   0
4    44.0  37.0   -7.0 -14.0  30.0       A1   1   0   0   0
..    ...   ...    ...   ...   ...      ...  ..  ..  ..  ..
195  63.0   3.0  -30.0 -24.0 -59.0       A1   1   0   0   0
196  97.0 -48.0  -61.0 -25.0 -21.0       B1   0   0   1   0
197 -93.0 -75.0  -18.0 -67.0 -58.0       B1   0   0   1   0
198  54.0 -66.0  -80.0  92.0  62.0       A1   1   0   0   0
199  82.0  53.0  -77.0  79.0  97.0       B2   0   0   0   1

[200 rows x 10 columns]

4. Convert the "one” column into 3 bins. (20 points)

In [1104]:
dummy_data.one = pd.cut(data.one, 3)

In [1105]:
dummy_data

one   two  three  four  five variable  A1  A2  B1  B2
0    (-209.667, -56.333] -76.0  -33.0   3.0 -13.0       B2   0   0   0   1
1        (-56.333, 97.0]  76.0   38.0  -6.0  80.0       B1   0   0   1   0
2        (-56.333, 97.0] -47.0  -34.0 -86.0 -66.0       A1   1   0   0   0
3    (-209.667, -56.333]  43.0    7.0 -40.0 -42.0       A1   1   0   0   0
4        (-56.333, 97.0]  37.0   -7.0 -14.0  30.0       A1   1   0   0   0
..                   ...   ...    ...   ...   ...      ...  ..  ..  ..  ..
195      (-56.333, 97.0]   3.0  -30.0 -24.0 -59.0       A1   1   0   0   0
196      (-56.333, 97.0] -48.0  -61.0 -25.0 -21.0       B1   0   0   1   0
197  (-209.667, -56.333] -75.0  -18.0 -67.0 -58.0       B1   0   0   1   0
198      (-56.333, 97.0] -66.0  -80.0  92.0  62.0       A1   1   0   0   0
199      (-56.333, 97.0]  53.0  -77.0  79.0  97.0       B2   0   0   0   1

[200 rows x 10 columns]

## Question 2

**Use the following speech by the Rev. Dr. Martin Luther King, Jr.:**

In [1106]:
s = "I am happy to join with you today in what will go down in history as the greatest demonstration for freedom in the history of our nation. Five score years ago, a great American, in whose symbolic shadow we stand today, signed the Emancipation Proclamation. This momentous decree came as a great beacon light of hope to millions of Negro slaves who had been seared in the flames of withering injustice. It came as a joyous daybreak to end the long night of their captivity. But one hundred years later, the Negro still is not free. One hundred years later, the life of the Negro is still sadly crippled by the manacles of segregation and the chains of discrimination. One hundred years later, the Negro lives on a lonely island of poverty in the midst of a vast ocean of material prosperity. One hundred years later, the Negro is still languishing in the corners of American society and finds himself an exile in his own land. So we have come here today to dramatize a shameful condition."

1. Find out how many unique words in s. (10 points)

In [1107]:
s = s.replace('.', '')
s = s.replace(',', '')
s = s.lower()

In [1108]:
s

'i am happy to join with you today in what will go down in history as the greatest demonstration for freedom in the history of our nation five score years ago a great american in whose symbolic shadow we stand today signed the emancipation proclamation this momentous decree came as a great beacon light of hope to millions of negro slaves who had been seared in the flames of withering injustice it came as a joyous daybreak to end the long night of their captivity but one hundred years later the negro still is not free one hundred years later the life of the negro is still sadly crippled by the manacles of segregation and the chains of discrimination one hundred years later the negro lives on a lonely island of poverty in the midst of a vast ocean of material prosperity one hundred years later the negro is still languishing in the corners of american society and finds himself an exile in his own land so we have come here today to dramatize a shameful condition'

In [1109]:
def unique_words(text):
    """Returns number of unique words in a string"""
    words = text.split(' ')
    return len(set(words))

In [1110]:
unique_words(s)

107

To find the number of unique words in the paragraph, I defined a function `unique_words()` that uses the `split()` method to convert the string to a list, used `set()` to get the set of distinct values in the list, then returned the set of unique word's length. Using the function on the raw string caused some words to be duplicated because of punctuation and case sensitivity. As a result, I removed any punctuation and made all characters lowercase using the `replace()` and `lower()` methods, respectively. My `unique_words()` returned all unique words and the sum of words which is 107.

2. Which word appears the most? (10 points)

In [1111]:
def most_freq(text):
    """ Returns the most frequent word in a string"""
    unique = list(set(text.split(' ')))
    counts = {i:text.count(i) for i in unique}
    return (max(counts, key=counts.get), max(counts.values()))

In [1112]:
most_freq(s)

('a', 70)

To find the most frequent word in the paragraph, I defined a function `most_freq()` that gets all unique words using the same method as the `unique_words()` function. Then, I used dictionary comprehension to get the count of each unique word. Finally, I used the built-in `max()` to return the most frequent word and its count. The function returned the word 'a' which appeared 70 times.

3. How many words start with ‘t’. (10 points).

In [1113]:
def start_with_t(str):
    """Returns the count of words that start with t in a string"""
    unique_t_count = 0
    para_t_count = 0
    unique = list(set(str.split(' ')))
    words = str.split(' ')
    for i in unique:
        if i.startswith('t'):
            unique_t_count += 1
    for j in words:
        if j.startswith('t'):
            para_t_count += 1
    return (unique_t_count, para_t_count)

In [1114]:
start_with_t(s)

(5, 23)

My final function `start_with_t()` take a string as input and returns the amount of unique words that starts with 't' and how many total words that appear in the string start with 't'. This function utilizes the built-in string method `startswith()`. My function return the tuple `(5, 23)`, meaning there are 5 unique words that start with 't', and those five words appeared 23 times total within the sentence.